File: my-folder/0402-remove-k-digits/solution.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        stack = []
        removed = 0
    
        for digit in num:
            while stack and removed < k and stack[-1] > digit:
                stack.pop()
                removed += 1
            stack.append(digit)
    
        while removed < k:
            stack.pop()
            removed += 1
    
        result = ''.join(stack).lstrip('0')
    
        return result if result else '0'

File: my-folder/0402-remove-k-digits/solution.java
class Solution {
    public String removeKdigits(String num, int k) {
        Stack<Character> stack = new Stack<>();
        for (char number : num.toCharArray()) {
            while (!stack.isEmpty() && stack.peek() > number && k > 0) {
                stack.pop();
                k--;
            }
            stack.push(number);
        }
        while (k > 0) {
            stack.pop();
            k--;
        }
        ArrayList<Character> res = new ArrayList<>(stack);
        while (!res.isEmpty() && res.get(0) == '0') {
            res.remove(0);
        }
        StringBuilder sb = new StringBuilder();
        for (char c : res) {
            sb.append(c);
        }
        String ans = sb.toString();
        if (ans.equals("")) {
            return "0";
        }
        return ans;
    }
}
